### PR TITLE
[bdwgc] New port

### PR DESCRIPTION
**Pull request by NNemec**
After having finished this new port, I found that there were already two issues requesting bdwgc pointing to existing branches on other repos (#5471 and #6052). Feel free to pick either one of those or this one.

### CODE REVIEW

**Comment by NancyLi1013** [comment]
Hi @NNemec, thanks for the new port. Here are the test results from the current CI system:

|x64-windows-static|master| test |notes
|------------------|------|------|-----
|bdwgc             |      |Fail  |New Port

|x64-osx|master| test |notes
|-------|------|------|-----
|bdwgc  |      |Fail  |New Port

|x64-linux|master| test |notes
|---------|------|------|-----
|bdwgc    |      |Fail  |New Port

|x64-uwp|master| test |notes
|-------|------|------|-----
|bdwgc  |      |Fail  |New Port

|arm-uwp|master| test |notes
|-------|------|------|-----
|bdwgc  |      |Fail  |New Port

[failureLogs.zip](https://github.com/microsoft/vcpkg/files/3170942/failureLogs.zip)

**Comment by vicroms** [comment]
As of right now, the port builds successfully on Windows and Linux. 
I haven't investigated what causes it to fail on UWP and MacOS X.

I'm a bit sceptical that the Linux build works properly, so I'll do some manual testing when I have time.

**Comment by NNemec** [comment]
Great! I got UWP building locally once and will try to integrate it within the next days.

File: ports/bdwgc/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+# LIB
+if (VCPKG_CRT_LINKAGE STREQUAL static)
+    set(LIBNAME "gcmt-lib${STATIC_LIB_SUFFIX}")
+else()
+    set(LIBNAME "gcmt-dll${STATIC_LIB_SUFFIX}")
+endif()
+
+if (NOT VCPKG_CMAKE_SYSTEM_NAME OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
+    file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/${LIBNAME}" DESTINATION ${CURRENT_PACKAGES_DIR}/lib RENAME gc${STATIC_LIB_SUFFIX})
```
Comment (ras0219-msft) [inline ports/bdwgc/portfile.cmake]: We should use the upstream names here as much as possible to maximize compatibility.

**Comment by NNemec** [comment]
I introduced the rename to gc.lib because that was what an existing FindGC.cmake was searching for. I'll have another close look at this with my current understanding.

**Comment by vicroms** [comment]
@NNemec 

I have local changes to the CMakeLists.txt to produce binaries with the names `gc.lib` and `gc.dll`.

I couldn't find a `FindGC.cmake` file in the source code, and there's also no FindGC.cmake module that ships with CMake. Is that file specific to your project?

I also noticed that when building on Linux (using the steps on their README), you get `libgc.so` and `libgc.a` binaries. So that leads me to believe that `<prefix>gc<suffix>` is the correct naming convention for the library.

Can I ask, how did you build the binaries before?
Are the names produced different than the ones produced by CMake?

**Comment by NNemec** [comment]
The FindGC.cmake is checked in to the project I'm working on. In no way authoritative, just one example of what is out there...

I never built GC before the approach I submitted here.

**Comment by vicroms** [comment]
I modified the CMakeLists.txt file to produce binaries with the `gc` name. 
These names should match the ones in the `debian` package for bdwgc.

**Comment by vicroms** [comment]
@NNemec 

Whenever you have time can you test the latest commit?
I think that your FindGC.cmake file should work with the added patch `001-install-libraries.patch`.

I still haven't gotten `bdwgc` to work with UPW or on Mac OS.
But as an initial port, Windows and Linux support look OK.

**Comment by NNemec** [comment]
Before I can try this out, I will first have to sort out some stuff in my local workspace. I will try over the weekend.

To get UWP working, the most straightforward approach is to deactivate the build of 'cord' - with your changes that should simply mean to drop BDWGC_INSTALL_TOOLS from the portfile.cmake.

As I understand it, cord is in no way essential for the libraries. It seems to be an interactive text editor using an advanced string class exploiting the garbage collector in some clever way. I doubt that anyone uses that code and I really see no reason to struggle with windows UI libraries when the goal is to get a low-level library compiled.

**Comment by NNemec** [comment]
Trying out the latest changes, I find that the include files are now installed as include/bdwgc/gc.h - why was that extra subdirectory introduced? The FindGC.cmake the the project that I am porting searches for include/gc.h which worked before, but now it breaks. Trivial to fix on either side, but it should be clarified what the proper way is.

**Comment by vicroms** [comment]
I'm modeling the port after the distributed Debian package, so the correct thing to do would be to remove the `bdwgc` folder and install directly to `include`, and that would match your `FindGC.cmake`.

**Comment by vicroms** [comment]
Submitted PR upstream (https://github.com/ivmai/bdwgc/pull/281).

With the hope to make these changes official if they make sense.

**Comment by vicroms** [comment]
As it stands, the current package works for all supported triplets.

I'm inclined to merging it right now, and update the port once the patches have been merged upstream.

@NNemec have you had time to test whether this port works for your project?

**Comment by NNemec** [comment]
My original submission works fine, I have not tried out any of the later versions and won't have a chance to do so for the next two weeks. Go ahead with merging. If I encounter any problems, it can still be fixed...

**Pull request merged**